Support IN16B's single detectors in IndirectILLEnergyTransfer

MathieuTillet opened the pull request:
**Description of work.**
In BATS mode, IN16B can have single detectors, that needs the same treatments as the other detectors, but without grouping.
This PR takes care of this, and fixes an indexation bug for regular pixel grouping.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**
- Run
 ```
center, epp = IndirectILLEnergyTransfer(Run='215962', PulseChopper='34', GroupDetectors=False)

offset_p100 = IndirectILLEnergyTransfer(Run='215967', InputElasticChannelWorkspace=epp, PulseChopper='34', GroupDetectors=False)
```
(with data from `ExternalData/Testing/SystemTest/ILL/IN16B`)
- Plot the color fill of the results and check there is an elastic peak neatly aligned.
- Plot spectra 2050 and 2051, and check the elastic peak is correctly centered on 0. 

WIP for #30591 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

MathieuTillet committed:
Fix grouping numbering bug

MathieuTillet committed:
Fix grouping numbering bug

MathieuTillet committed:
Compute energy transfer for single detectors

IN16B has single detectors, which were ignored up to this point when
computing the energy transfer and the epp. They are now taken into
account.

MathieuTillet committed:
Compute energy transfer for single detectors

IN16B has single detectors, which were ignored up to this point when
computing the energy transfer and the epp. They are now taken into
account.

MathieuTillet committed:
Convert monitor to TOF

MathieuTillet committed:
Convert monitor to TOF

MathieuTillet committed:
Update relevant docs

MathieuTillet committed:
Update relevant docs

MathieuTillet committed:
Set default value for elastic tof and channel

MathieuTillet committed:
Set default value for elastic tof and channel

MathieuTillet committed:
Remove NarrowPeak from peak detection

Stop considering that a peak fitted with a NarrowPeak flag is valid, and
instead default to the equatorial peak. this is because this NarrowPeak
has been found to produce less accurate results.

MathieuTillet committed:
Remove NarrowPeak from peak detection

Stop considering that a peak fitted with a NarrowPeak flag is valid, and
instead default to the equatorial peak. this is because this NarrowPeak
has been found to produce less accurate results.

MathieuTillet committed:
Modify reference data for system test

MathieuTillet committed:
Modify reference data for system test

MathieuTillet committed:
Merge branch 'master' into 30591_in16b_single_detectors

MathieuTillet committed:
Add release notes

MathieuTillet committed:
Add release notes

MathieuTillet committed:
Add t0 offset for SD

Compute and use t0 offset for single detectors, which can be different
from psd offset.

MathieuTillet committed:
Add t0 offset for SD

Compute and use t0 offset for single detectors, which can be different
from psd offset.

MathieuTillet committed:
Update systemtest reference

MathieuTillet committed:
Update systemtest reference

MathieuTillet committed:
Merge branch 'master' into 30591_in16b_single_detectors

MathieuTillet committed:
Merge branch 'master' into 30591_in16b_single_detectors

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:
```diff
@@ -437,10 +442,14 @@ def _create_elastic_channel_ws(self, epp_ws, run, epp_equator_ws=None):
             epp_ws.setCell('ChopperPhase', row, phase)
-            epp_ws.setCell('PSD_TOF_Delay', row, delay)
+            if epp_ws.rowCount() - single_detectors - equator_epp <= row < epp_ws.rowCount() - equator_epp:
+                epp_ws.setCell('PSD_TOF_Delay', row, delay_sd)
```
Now the delay is both for PSD and SD, so let's rename it to just `TOF_Delay`.

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:
```diff
@@ -502,26 +512,44 @@ def _convert_to_energy_bats(self, ws, epp_ws, t0_offset=0.):
+            else:
+                elastic_tof = elastic_tof_equator
+                elastic_channel = elastic_channel_equator
+                self.log.warning("Single detector number {0}'s peak was not found. Using equatorial peak instead."
```
In this case could we not calculate a guess based on the relative distance and relative delay differences of the PSD equator and SD?

gvardany replied:
This feature is not urgent, so I created a separate issue to address it later: https://github.com/mantidproject/mantid/issues/31399

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:
```diff
@@ -592,26 +621,48 @@ def _reduce_bats(self, ws):
+                sd_t0_offset = self._t0_offset(center_chopper_speed, center_chopper_phase, shifted_chopper_phase,
+                                               center_sd_delay, shifted_sd_delay)
+
+            self.log().information('T0 Offset is {0} [sec]'.format(psd_t0_offset))
```
`PSD T0 Offset...`

gvardany commented on Testing/Data/SystemTest/ILL/IN16B/ILLIN16B_BATS.nxs.md5:
```diff
@@ -0,0 +1 @@
+5922f89b45b9f73becd769a2d23154e5
```
Is this file still needed?

gvardany commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py:
```diff
@@ -592,26 +621,48 @@ def _reduce_bats(self, ws):
+                               StartWorkspaceIndex=N_TUBES*N_PIXELS_PER_TUBE + 1,
+                               EndWorkspaceIndex=N_TUBES*N_PIXELS_PER_TUBE + single_detectors)
+                FindEPP(InputWorkspace=single_det_ws, OutputWorkspace=epp_ws)
+                self._create_elastic_channel_ws(mtd[epp_ws], mtd[ws].getRun(), mtd[equator_epp_ws],
```
This is a bit confusing, are you passing the epp of single detectors as the first argument, together with number of single detectors? If so, maybe clarify in the doxygen comments of _create_elastic...

gvardany reviewed (commented):
Few comments on the code, will check with the new data shortly.

gvardany reviewed (commented):
Tested with 2 new measurements, and it all worked as expected.
Just noticed an issue in case GroupDetectors=True.
This is not related to the grouping done for the peak finding.
This is the final grouping of the output.
The issue is that the default grouping file contains only 2 single detectors.
So if there are more, the rest are not included in the final result.
This cannot be added in the default grouping file, since in that case having 2 single detectors will fail due to unknown spectrum/detid for the 3rd. We have to think of something more dynamic but this can be done in a separate ticket. For this one, I'd only suggest to add one more system test with the new data, that has 3 single detectors (with GroupDetectors=False). The test should be similar to the existing one, i.e. use a center measurement and an offseted one.

StephenSmith25 commented:
Is it intended for this PR to make it into the 6.1 release?

MathieuTillet commented:
Yes, it should be ready soon.

MathieuTillet committed:
Update tests and test files

MathieuTillet committed:
Update tests and test files

MathieuTillet committed:
Correct pixel grouping

Takes into account monitors and the 3 single detectors case

MathieuTillet committed:
Correct pixel grouping

Takes into account monitors and the 3 single detectors case

MathieuTillet committed:
Dynamically compute grouping pattern

In IndirectILLEnergyTransfer, compute the grouping pattern on the fly
rather than relying on grouping files.

MathieuTillet committed:
Dynamically compute grouping pattern

In IndirectILLEnergyTransfer, compute the grouping pattern on the fly
rather than relying on grouping files.

MathieuTillet committed:
Fix equatorial fitting issue

There was an offset by one. Added test to check that too.

MathieuTillet committed:
Fix equatorial fitting issue

There was an offset by one. Added test to check that too.

gvardany commented:
Almost there. There is an issue still with FitEquatorialOnly, when having also an offset measurement:
```
config.appendDataSearchDir('/Volumes/illdata/211/in16b/internalUse/rawdata')
center, epp = IndirectILLEnergyTransfer(Run='318308', 
                                        PulseChopper='34',
                                        ElasticPeakFitting='FitEquatorialOnly',  
                                        GroupDetectors=False)
offset = IndirectILLEnergyTransfer(Run='315515:315521', 
                                   PulseChopper='34', 
                                   GroupDetectors=False,
                                   InputElasticChannelWorkspace=epp)
```
Take a look into this and turn this to a system test to cover also the offset measurement where this EPP table is actually used.
Ideally add 2 tests, one for each ElasticPeakFitting method.

MathieuTillet committed:
Fix issue with reusing fit data and add test

ILLIndirectEnergyTransfer would crash if using epp computed with
FitEquatorial only on a FitAllPixels option.

MathieuTillet committed:
Fix issue with reusing fit data and add test

ILLIndirectEnergyTransfer would crash if using epp computed with
FitEquatorial only on a FitAllPixels option.

gvardany reviewed (dismissed):
Thanks for adding the test, looks good.

gemmaguest commented:
@MathieuTillet Could you change the base branch to `release-next` if this is for the release?

MathieuTillet committed:
Merge branch '30591_in16b_single_detectors' of github.com:mantidproject/mantid into 30591_in16b_single_detectors
